File: OPF_Matlab/opf_dc_cvx.m
clear all; clc;

testsystem = "case_ieee30";
load("data/network/mat/" + testsystem + ".mat"); % load function from MATPOWER

%%
[buses, lines, generators] = readnetwork(mpc);


lineset = 1:size( lines, 1 );
busset = 1:size( buses, 1 );
genset = 1:size( generators, 1);

B_g = [];
for g = genset
    B_g = [ B_g; generators(g).location ];
end

%%
B_gn_cell = cell( size( buses, 1 ), 1 );
for g = genset
    B_gn_cell{generators(g).location} = [B_gn_cell{generators(g).location}; g];
end
B_gn = containers.Map( busset, B_gn_cell);

[ Ybus, yff, yft, ytf, ytt ] = ybus( buses, lines );
yff_r = real(yff);
yff_i = imag(yff);
ytt_r = real(ytt);
ytt_i = imag(ytt);
yft_r = real(yft);
yft_i = imag(yft);
ytf_r = real(ytf);
ytf_i = imag(ytf);

%%
gencost = zeros(length(genset),3);
for g = genset
    gencost(g,:) = generators(g).cost;
end

cvx_solver Gurobi
% cvx_solver Knitro
% cvx_solver Mosek

% Create and solve problem
cvx_begin
variable theta( length(busset), 1 );
variable pg( length(genset), 1 );
variable p_ft( length(lineset), 1 );
variable p_tf( length(lineset), 1 );

OCgen = ( gencost(:,1) .* (pg * mpc.baseMVA).^2 ...
    + gencost(:,2) .* (pg * mpc.baseMVA) ...
    + gencost(:,3) );
OC = sum( OCgen );

minimize( OC )
subject to
for b = busset
    - deg2rad(180) <= theta(b) <= deg2rad(180)
end
for g = genset
    generators(g).Pmin <= pg(g) <= generators(g).Pmax;
end

for b = busset
    if buses(b).btype == 3
        theta(b) == 0;
    end
end
for l = lineset
    p_ft(l) == yft_i(l) * ( theta( lines(l).fbus ) - theta( lines(l).tbus ) );
    p_tf(l) == ytf_i(l) * ( theta( lines(l).tbus ) - theta( lines(l).fbus ) );
    
    if lines(l).u ~= 0
        - lines(l).u <= p_ft(l) <= lines(l).u;
        - lines(l).u <= p_tf(l) <= lines(l).u;
    end
    
    lines(l).angmin <= theta(lines(l).fbus) - theta(lines(l).tbus) <= lines(l).angmax;
    % max(lines(l).angmin, -deg2rad(60)) <= theta(lines(l).fbus) - theta(lines(l).tbus) <= min(lines(l).angmax, deg2rad(60));
end
for b = busset
    sum( p_ft( buses(b).outline ) ) ...
        + sum( p_tf( buses(b).inline ) ) ...
        - sum( pg( B_gn(b) )) ...
        + buses(b).Pd ...
        + buses(b).Gs * 1.0^2 ...
        == 0;
end
cvx_end

%% Run the optimization
cvx_objvalue = value( OC );
%%

File: OPF_Python/opf_dc_cvx.py
#%%
import numpy as np
import cvxpy as cp

# from dataclasses import dataclass, field
# from typing import List
from math import radians

#%%
# from src.struct_network import Bus, Line, Generator
from src.readnetwork import readnetwork
from src.ybus import ybus
#%%


testsystem = "case_ieee30"

#%%
buses, lines, generators, datamat = readnetwork(testsystem)

nline = len(lines)
nbus = len(buses)
ngen = len(generators)
lineset = np.arange( 0, nline )
busset = np.arange( 0, nbus )
genset = np.arange(0, ngen )
baseMVA = datamat["baseMVA"][0, 0][0, 0]

B_g = []
for g in genset:
    B_g.append( generators[g].location )

B_gn = []
for i in busset:
    B_gn.append( [] )

for g in genset:
    B_gn[ generators[g].location ].append( int(g) )


#%%
Ybus, yff, yft, ytf, ytt = ybus( buses, lines )

yff_r = np.real(yff)
yff_i = np.imag(yff)
ytt_r = np.real(ytt)
ytt_i = np.imag(ytt)
yft_r = np.real(yft)
yft_i = np.imag(yft)
ytf_r = np.real(ytf)
ytf_i = np.imag(ytf)



#%% cvx start

theta = cp.Variable( len(busset))
pg = cp.Variable( len(genset) )
p_ft = cp.Variable( len(lineset) )
p_tf = cp.Variable( len(lineset) )

constraints = []
for b in busset:
    constraints += [ - np.pi <= theta[b] ]
    constraints += [ theta[b] <= + np.pi ] 
    if buses[b].btype == 3:
        constraints += [ theta[b] == 0]
    
for g in genset:
    constraints += [ generators[g].Pmin <= pg[g] ]
    constraints += [ pg[g] <= generators[g].Pmax ]

for l in lineset:
    constraints += [ 
        p_ft[l] == float(yft_i[l]) * (theta[lines[l].fbus]-theta[lines[l].tbus]) 
            ]
    constraints += [ 
        p_tf[l] == float(ytf_i[l]) * (theta[lines[l].tbus]-theta[lines[l].fbus])
            ]


for b in busset:
    constraints += [ 
        sum(p_ft[l] for l in buses[b].outline) 
        + sum(p_tf[l] for l in buses[b].inline) 
        - sum(pg[g] for g in B_gn[b]) 
        + buses[b].Pd 
        + buses[b].Gs * 1.0 ** 2 
        == 0
        ]

for l in lineset:
    if lines[l].u != 0:
        constraints += [ p_ft[l] <= + lines[l].u ]
        constraints += [ p_ft[l] >= - lines[l].u ]
        constraints += [ p_tf[l] <= + lines[l].u ]
        constraints += [ p_tf[l] >= - lines[l].u ]    
    constraints += [ theta[lines[l].fbus] - theta[lines[l].tbus] <= lines[l].angmax ]
    constraints += [ theta[lines[l].fbus] - theta[lines[l].tbus] >= lines[l].angmin ]
    # constraints += [ theta[lines[l].fbus] - theta[lines[l].tbus] <= min(lines[l].angmax, radians(60)) ]
    # constraints += [ theta[lines[l].fbus] - theta[lines[l].tbus] >= max(lines[l].angmin, -radians(60)) ]

prob = cp.Problem(cp.Minimize(
    sum(
        (generators[g].cost[0] * (pg[g]*baseMVA)**2
        + generators[g].cost[1] * (pg[g]*baseMVA) 
        + generators[g].cost[2]) 
            for g in genset) ), constraints)


#%% 
# print(cp.installed_solvers())
prob.solve( solver=cp.GUROBI , verbose = True )
# prob.solve( solver=cp.MOSEK , verbose = True )
# prob.solve( verbose = True )
print("Termination status =", prob.status)
print("The optimal value is", prob.value)
print("test system =", testsystem)
print("solve time =", prob.solver_stats.solve_time)
